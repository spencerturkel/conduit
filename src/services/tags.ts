import {Observable} from 'rxjs/Observable';
import {ajax} from 'rxjs/observable/dom/ajax';
import {map} from 'rxjs/operators';

export const tags$: Observable<string[]> = ajax({
    url: 'https://conduit.productionready.io/api/tags',
    async: true,
    method: 'GET',
    responseType: 'json',
}).pipe(map(({response: {tags}}) => tags));
