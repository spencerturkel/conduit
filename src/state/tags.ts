import {Observable} from 'rxjs/Observable';
import {first, map, switchMap} from 'rxjs/operators';

export enum ActionType {
    FETCH_ALL = '[conduit][tags] Fetch All',
    FETCH_ALL_ERROR = '[conduit][tags] Fetch All Error',
    LOAD_ALL = '[conduit][tags] Load All',
}

export interface FetchAllAction {
    type: ActionType.FETCH_ALL;
}

export interface FetchAllErrorAction {
    error: any;
    type: ActionType.FETCH_ALL_ERROR;
}

export interface LoadAllAction {
    tags: string[];
    type: ActionType.LOAD_ALL;
}

export type Action = FetchAllAction | FetchAllErrorAction | LoadAllAction;

export const fetchAll = (): FetchAllAction => ({type: ActionType.FETCH_ALL});
export const fetchAllError = (error: any): FetchAllErrorAction => ({error, type: ActionType.FETCH_ALL_ERROR});
export const loadAll = (tags: string[]): LoadAllAction => ({tags, type: ActionType.LOAD_ALL});

export interface State {
    error?: any;
    loading: boolean;
    tags: string[];
}

export const initialState: State = {
    loading: false,
    tags: [],
};

export const reducer = (state: State | undefined = initialState, action: Action): State => {
    switch (action.type) {
        case ActionType.FETCH_ALL:
            return {...state, error: undefined, loading: true};
        case ActionType.FETCH_ALL_ERROR:
            return {...state, error: action.error, loading: false};
        case ActionType.LOAD_ALL:
            return {error: undefined, loading: false, tags: action.tags};
        default:
            return state;
    }
};

export const fetchAll$ = (tags$: () => Observable<string[]>) => (actions$: Observable<Action>): Observable<Action> =>
    actions$.pipe(first(({type}) => type === ActionType.FETCH_ALL), switchMap(() => tags$()), map(loadAll));

export const selectError = (state: State) => state.error;
export const selectLoading = (state: State) => state.loading;
export const selectTags = (state: State) => state.tags;
