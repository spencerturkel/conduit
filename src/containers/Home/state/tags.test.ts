import {TestScheduler} from 'rxjs/testing/TestScheduler';

import {fetchAllError, fetchAll, fetchAll$, loadAll, initialState, reducer, Action, State} from './tags';

const testStates = (transform: (initial: State | undefined) => any) => {
    expect(transform(undefined)).toMatchSnapshot();

    expect(transform({loading: true, tags: []})).toMatchSnapshot();

    expect(transform({loading: true, tags: ['hello', 'world']})).toMatchSnapshot();

    expect(transform({loading: false, tags: ['hello', 'world']})).toMatchSnapshot();

    expect(transform({error: 'some error', loading: false, tags: ['hello', 'world']})).toMatchSnapshot();
};

describe('reducers', () => {
    it('should return initial state given an unknown action and an undefined state', () => {
        const state = reducer(undefined, {type: Math.random().toString()} as Action);

        expect(state).toBe(initialState);
    });

    describe('action creators', () => {
        const actionYieldsExpectedState = (name: string, action: Action) =>
            it(`should match snapshot for ${name}()`, () => testStates(state => reducer(state, action)));

        actionYieldsExpectedState('fetchAll', fetchAll());
        actionYieldsExpectedState('fetchAllError', fetchAllError('test error'));
        actionYieldsExpectedState('loadAll', loadAll(['tag one', 'tag two', 'tag three']));
    });
});

describe('epics', () => {
    let scheduler: TestScheduler;

    beforeEach(() => {
        scheduler = new TestScheduler((actual, expected) => expect(actual).toEqual(expected));
    });

    describe(fetchAll$.name, () => {
        it('should emit LOAD_ALL actions after the first FETCH_ALL action', () => {
            const tags$ = scheduler.createColdObservable<string[]>('--a--b|', {a: [], b: ['hello', 'world']});

            const sut = fetchAll$(tags$);

            const actions$ = scheduler.createColdObservable<Action>('-a--b', {a: {type: 'unknown'}, b: fetchAll()});

            scheduler
                .expectObservable(sut(actions$))
                .toBe('------a--b', {a: loadAll([]), b: loadAll(['hello', 'world'])});
        });

        it('should catch and report errors in the tag source without completing', () => {
            const tags$ = scheduler.createColdObservable<string[]>('#', undefined, 'err');

            const sut = fetchAll$(tags$);

            const actions$ = scheduler.createColdObservable<Action>('ab', {a: fetchAll(), b: fetchAll()});

            scheduler.expectObservable(sut(actions$)).toBe('aa', {a: fetchAllError('err')});
        });
    });
});
