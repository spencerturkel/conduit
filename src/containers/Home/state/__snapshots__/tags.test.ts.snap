// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers action creators should match snapshot for fetchAll() 1`] = `
Object {
  "error": undefined,
  "loading": true,
  "tags": Array [],
}
`;

exports[`reducers action creators should match snapshot for fetchAll() 2`] = `
Object {
  "error": undefined,
  "loading": true,
  "tags": Array [],
}
`;

exports[`reducers action creators should match snapshot for fetchAll() 3`] = `
Object {
  "error": undefined,
  "loading": true,
  "tags": Array [
    "hello",
    "world",
  ],
}
`;

exports[`reducers action creators should match snapshot for fetchAll() 4`] = `
Object {
  "error": undefined,
  "loading": true,
  "tags": Array [
    "hello",
    "world",
  ],
}
`;

exports[`reducers action creators should match snapshot for fetchAll() 5`] = `
Object {
  "error": undefined,
  "loading": true,
  "tags": Array [
    "hello",
    "world",
  ],
}
`;

exports[`reducers action creators should match snapshot for fetchAllError() 1`] = `
Object {
  "error": "test error",
  "loading": false,
  "tags": Array [],
}
`;

exports[`reducers action creators should match snapshot for fetchAllError() 2`] = `
Object {
  "error": "test error",
  "loading": false,
  "tags": Array [],
}
`;

exports[`reducers action creators should match snapshot for fetchAllError() 3`] = `
Object {
  "error": "test error",
  "loading": false,
  "tags": Array [
    "hello",
    "world",
  ],
}
`;

exports[`reducers action creators should match snapshot for fetchAllError() 4`] = `
Object {
  "error": "test error",
  "loading": false,
  "tags": Array [
    "hello",
    "world",
  ],
}
`;

exports[`reducers action creators should match snapshot for fetchAllError() 5`] = `
Object {
  "error": "test error",
  "loading": false,
  "tags": Array [
    "hello",
    "world",
  ],
}
`;

exports[`reducers action creators should match snapshot for loadAll() 1`] = `
Object {
  "error": undefined,
  "loading": false,
  "tags": Array [
    "tag one",
    "tag two",
    "tag three",
  ],
}
`;

exports[`reducers action creators should match snapshot for loadAll() 2`] = `
Object {
  "error": undefined,
  "loading": false,
  "tags": Array [
    "tag one",
    "tag two",
    "tag three",
  ],
}
`;

exports[`reducers action creators should match snapshot for loadAll() 3`] = `
Object {
  "error": undefined,
  "loading": false,
  "tags": Array [
    "tag one",
    "tag two",
    "tag three",
  ],
}
`;

exports[`reducers action creators should match snapshot for loadAll() 4`] = `
Object {
  "error": undefined,
  "loading": false,
  "tags": Array [
    "tag one",
    "tag two",
    "tag three",
  ],
}
`;

exports[`reducers action creators should match snapshot for loadAll() 5`] = `
Object {
  "error": undefined,
  "loading": false,
  "tags": Array [
    "tag one",
    "tag two",
    "tag three",
  ],
}
`;
